File: SourceCode/array/0581_Shortest_Unsorted_Continuous_Subarray.py
class Solution(object):
    def findUnsortedSubarray(self, nums):
        n = len(nums)
        maxn, r = float("-inf"), -1
        minn, l = float("inf"), -1
        for i in range(n):
            if minn >= nums[n-i-1]:
                minn = nums[n-i-1]
            else:
                l = n - i - 1

            if maxn <= nums[i]:
                maxn = nums[i]
            else:
                r = i
        return 0 if r == -1 else r - l + 1

File: SourceCode/MonotonousStack/0581_Shortest_Unsorted_Continuous_Subarray.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
  int findUnsortedSubarray(vector<int>& nums) {
    stack<int> sta;
    int n = nums.size();
    int l = n, r = 0;
    for(int i = 0; i < n; i++){
      while(!sta.empty() && nums[i] < nums[sta.top()]) {
        l = min(l, sta.top());
        sta.pop();
      }
      sta.push(i);
    }
    sta = stack<int>();
    for(int i = n-1; i >=0; i--){
      while(!sta.empty() && nums[i] > nums[sta.top()]) {
        r = max(r, sta.top());
        sta.pop();
      }
      sta.push(i);
    }
    return (r-l) < 0 ? 0 : r - l + 1;
  }
};